src/librustc_typeck/collect.rs

```diff
@@ -2634,6 +2636,16 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
     codegen_fn_attrs
 }
 
+fn inline_exportable(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+    // Functions marked with #[inline] are only ever codegened
+    // with "internal" linkage and are never exported unless we're
+    // building a `staticlib` or `cdylib` and they are marked
+    // `#[no_mangle]`.
+    tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::NO_MANGLE)
+        && (tcx.sess.crate_types().contains(&CrateType::Cdylib)
+            || tcx.sess.crate_types().contains(&CrateType::Staticlib))
```

Comment:
Do we need to check the crate type here? Even when building `rlib`s we'll still need to codegen and export `#[no_mangle] #[inline]` items in case they're linked into a `staticlib` or `cdylib`.

Reply:
Just to check I understand between these two you're saying that `#[inline]` should always be superseded by an extern indicator and crate-type/specifics of indicator are irrelevant?

Reply:
Yes, this should be done regardless of the output crate type.